cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +200,65 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source final : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<std::byte const> d_buffer) : _d_buffer{d_buffer}
+  {
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpy(dst, _d_buffer.data() + offset, count, cudaMemcpyDefault));
+    return count;
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    std::vector<std::byte> h_data(count);
+    CUDF_CUDA_TRY(cudaMemcpy(h_data.data(), _d_buffer.data() + offset, count, cudaMemcpyDefault));
+    return std::make_unique<owning_buffer<std::vector<std::byte>>>(std::move(h_data));
+  }
+
+  [[nodiscard]] bool supports_device_read() const override { return true; }
+
+  std::future<size_t> device_read_async(size_t offset,
+                                        size_t size,
+                                        uint8_t* dst,
```

Comment:
I see that we are using mixed `uint8_t`, `char` and `std::byte`. Can we just use one standard type like `std::byte` or `char` (out-of-scope of this PR)?

Reply:
Agreed. We should unify all byte types by using `std::byte` consistently. cc @vuule

Reply:
yes. we should be using std::byte for "raw data" buffers (https://github.com/rapidsai/cudf/issues/6029#issuecomment-1302662162).
It's a big task and I haven't decided on how best to approach it.
If there's interest, I can open an issue and we can discuss there.

Reply:
+1  for using `std::byte`